jupyterhub/oauth/store.py

```diff
@@ -0,0 +1,315 @@
+        handler = request.handler
+        user = handler.get_current_user()
+        if user:
+            return {}, user.id
+        else:
+            raise UserNotAuthenticated()
+
+    def user_has_denied_access(self, request):
+        # user can't deny access
+        return False
+    
+
+class HubDBMixin(object):
+    """Mixin for connecting to the hub database"""
+    def __init__(self, session_factory):
+        self.session_factory = session_factory
+        self._local = threading.local()
+
+    @property
+    def db(self):
+        if not hasattr(self._local, 'db'):
+            self._local.db = scoped_session(self.session_factory)()
+        return self._local.db
+
+
+class AccessTokenStore(HubDBMixin, oauth2.store.AccessTokenStore):
+    """OAuth2 AccessTokenStore, storing data in the Hub database"""
+
+    def _access_token_from_orm(self, orm_token):
+        """Transform an ORM AccessToken record into an oauth2 AccessToken instance"""
+        return AccessToken(
+            client_id=orm_token.client_id,
+            grant_type=orm_token.grant_type,
+            expires_at=orm_token.expires_at,
+            refresh_token=orm_token.refresh_token,
+            refresh_expires_at=orm_token.refresh_expires_at,
+            user_id=orm_token.user_id,
+        )
+
+    def save_token(self, access_token):
+        """
+        Stores an access token in the database.
+
+        :param access_token: An instance of :class:`oauth2.datatype.AccessToken`.
+
+        """
+        
+        user = self.db.query(orm.User).filter(orm.User.id == access_token.user_id).first()
+        token = user.new_api_token(access_token.token)
+        orm_api_token = APIToken.find(self.db, token, kind='user')
+        
+        orm_access_token = orm.OAuthAccessToken(
+            client_id=access_token.client_id,
+            grant_type=access_token.grant_type,
+            expires_at=access_token.expires_at,
+            refresh_token=access_token.refresh_token,
+            refresh_expires_at=access_token.refresh_expires_at,
+            user=user,
+            api_token=orm_api_token,
+        )
+        self.db.add(orm_access_token)
+        self.db.commit()
+
+    def fetch_existing_token_of_user(self, client_id, grant_type, user_id):
+        """
+        Fetches an access token identified by its client id, type of grant and
+        user id.
+
+        This method must be implemented to make use of unique access tokens.
+
+        :param client_id: Identifier of the client a token belongs to.
+        :param grant_type: The type of the grant that created the token
+        :param user_id: Identifier of the user a token belongs to.
+        :return: An instance of :class:`oauth2.datatype.AccessToken`.
+        :raises: :class:`oauth2.error.AccessTokenNotFound` if no data could be
+                 retrieved.
+        """
+        raise NotImplementedError("Unique tokens not implemented")
```

Comment:
Yup, the AccessTokenStore API defines several methods that our subset of OAuth doesn't need or use, so I haven't implemented them in our implementation. I've now removed them instead of leaving the NotImplementedErrors.